{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c12d6c64-2da0-475f-ad51-c1ba00924a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import subprocess\n",
    "from datetime import datetime\n",
    "import json\n",
    "import baltic as bt\n",
    "import caffeine\n",
    "\n",
    "module_dir = \"/Users/monclalab1/Documents/scripts/\"\n",
    "sys.path.append(module_dir)\n",
    "\n",
    "from fasta_editing import fasta_to_df, fasta_writer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3b5f55c5-c69f-4999-851d-ae075ca5ee52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#treetime needs a dates.csv file to root the tree\n",
    "#the date column needs to be in decimal format\n",
    "\n",
    "def create_date_csv(strain_date_map, output_dir):\n",
    "    \n",
    "    file_name = f\"{output_dir}/strain_dates.csv\"\n",
    "    \n",
    "    with open(file_name, 'w') as date_file:\n",
    "        date_file.write(\"name, date\\n\")\n",
    "        \n",
    "        for strain, date_str in strain_date_map.items():\n",
    "\n",
    "            date = datetime.strptime(date_str, '%Y-%m-%d')\n",
    "            \n",
    "            # convert to decimal year because thats what treetime uses\n",
    "            dec_date = date.year + ((date.month - 1) * 30 + date.day) / 365.0\n",
    "        \n",
    "            date_file.write(f\"{strain}, {dec_date:.2f}\\n\")\n",
    "    \n",
    "    return file_name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7b74a4cb-c7da-4734-92f1-da22906c291c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def treesort_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "                  newick_path, output_dir, ref_gene, reroot=False):\n",
    "    \n",
    "    descriptor_entries = [] \n",
    "\n",
    "    for gene in list_of_genes:\n",
    "\n",
    "        align_file = f\"{home_folder}/{align_path}/h3nx_{gene}.fasta\"\n",
    "        metadata_file = f\"{home_folder}/{metadata_path}/metadata_h3nx_{gene}.tsv\"\n",
    "        nwk_file = f\"{home_folder}/{newick_path}/h3nx_{gene}.nwk\"\n",
    "        output_align = f\"{output_dir}/alignments/h3nx_{gene}.fasta\"\n",
    "        output_metadata = f\"{output_dir}/metadata/h3nx_{gene}.csv\"\n",
    "        output_nwk = f\"{output_dir}/div_tree/h3nx_{gene}.nwk\"\n",
    "        \n",
    "        for path in [f\"{output_dir}/alignments\", f\"{output_dir}/metadata\", f\"{output_dir}/div_tree\"]:\n",
    "            if not os.path.exists(path):\n",
    "                os.makedirs(path)\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "        # metadata file\n",
    "\n",
    "        metadata = pd.read_csv(metadata_file, sep='\\t')\n",
    "        metadata[\"date\"] = metadata[\"date\"].str.replace('XX', '01')\n",
    "        \n",
    "        # alignment file \n",
    "        align = fasta_to_df(align_file)\n",
    "        align.header = align.header.str.replace(\">\", \"\")\n",
    "\n",
    "        merged = pd.merge(align, metadata[['strain', 'date']], left_on='header', right_on='strain', how='left')\n",
    "        merged[\"header\"] = merged[['strain', 'date']].apply('|'.join, axis=1)\n",
    "        merged[\"header\"] = \">\" + merged[\"header\"]\n",
    "        fasta_writer(f\"{output_dir}/alignments/\", f\"h3nx_{gene}.fasta\", merged)\n",
    "       \n",
    "        metadata[\"strain\"] = metadata[['strain', 'date']].apply('|'.join, axis=1)\n",
    "        metadata.to_csv(output_metadata, index=False)\n",
    "\n",
    "        # stain date map for treetime\n",
    "        strain_date_map = dict(zip(merged['strain'].str.replace(\">\", \"\"), merged['date']))\n",
    "        # strain|date date map for the tree file\n",
    "        aln_date_map = dict(zip(merged['header'].str.replace(\">\", \"\"), merged['date']))\n",
    "\n",
    "        # tree file\n",
    "        \n",
    "        def replace_strain(match):\n",
    "            strain = match.group(0)[:-1]  # remove trailing ':'\n",
    "            if strain in strain_date_map:\n",
    "                return f\"{strain}|{strain_date_map[strain]}:\"\n",
    "            return match.group(0)\n",
    "        \n",
    "        with open(nwk_file, 'r') as file:\n",
    "            nwk_data = file.read()\n",
    "            \n",
    "        strain_pattern = re.compile(r'\\bA[^:]+:')\n",
    "        updated_nwk_data = strain_pattern.sub(replace_strain, nwk_data)\n",
    "        \n",
    "        with open(output_nwk, 'w') as file:\n",
    "            file.write(updated_nwk_data)\n",
    "\n",
    "\n",
    "        gene_label = f\"*{gene.upper()}\" if gene == ref_gene else gene.upper()\n",
    "\n",
    "        # rerooting using treetime if needed \n",
    "        if gene == ref_gene and reroot:\n",
    "            create_date_csv(aln_date_map, output_dir)\n",
    "            subprocess.run([\n",
    "                \"treetime\", \"clock\",\n",
    "                \"--tree\", output_nwk,\n",
    "                \"--dates\", f\"{output_dir}/strain_dates.csv\",\n",
    "                \"--aln\", output_align,\n",
    "                \"--outdir\", f\"{output_dir}/div_tree\"\n",
    "            ], check=True)\n",
    "            descriptor_entries.append([gene_label, output_align, f\"{output_dir}/div_tree/rerooted.newick\"])\n",
    "        else:\n",
    "            descriptor_entries.append([gene_label, output_align, output_nwk])\n",
    "\n",
    "    # descriptor csv\n",
    "    with open(\"descriptor.csv\", 'w') as descriptor_file:\n",
    "        for row in descriptor_entries:\n",
    "            descriptor_file.write(','.join(row) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd4d9cec-ef5b-401f-9a51-907b27ea5db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def wrap_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "              newick_path, output_dir, ref_gene, \n",
    "              reroot=True, treesort=True, mincut=False):\n",
    "    \n",
    "    \n",
    "    treesort_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "                  newick_path, output_dir, ref_gene, reroot = True)\n",
    "    \n",
    "    if treesort:\n",
    "        \n",
    "        for i in range(1,6):\n",
    "        \n",
    "            if not os.path.exists(f\"trees_{i}\"):\n",
    "                os.makedirs(f\"trees_{i}\")\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "            caffeine.on(display=False)\n",
    "\n",
    "            if mincut:\n",
    "\n",
    "                treesort_cmd = [\n",
    "                            \"treesort\",\n",
    "                            \"-i\", \"descriptor.csv\",\n",
    "                            \"-o\", f\"trees_{i}/annotated.tree\",\n",
    "                            \"-m\", \"mincut\"\n",
    "                        ]\n",
    "            else:\n",
    "\n",
    "                treesort_cmd = [\n",
    "                            \"treesort\",\n",
    "                            \"-i\", \"descriptor.csv\",\n",
    "                            \"-o\", f\"trees_{i}/annotated.tree\"\n",
    "                        ]\n",
    "            subprocess.run(treesort_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c1a42033-4fc4-4ea9-9356-533612d8d68b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "0.00\t-TreeAnc: set-up\n",
      "\n",
      "Attempting to parse dates...\n",
      "\tUsing column 'name' as name. This needs match the taxon names in the tree!!\n",
      "\tUsing column 'date' as date.\n",
      "\n",
      "0.36\tTreeTime.reroot: with method or node: least-squares\n",
      "\n",
      "0.36\tTreeTime.reroot: rerooting will ignore covariance and shared ancestry.\n",
      "\n",
      "0.70\tTreeTime.reroot: with method or node: least-squares\n",
      "\n",
      "0.71\tTreeTime.reroot: rerooting will ignore covariance and shared ancestry.\n",
      "\n",
      "1.03\tTreeTime.reroot: with method or node: least-squares\n",
      "\n",
      "1.03\tTreeTime.reroot: rerooting will ignore covariance and shared ancestry.\n",
      "\n",
      " Root-Tip-Regression:\n",
      " --rate:\t1.318e-03\n",
      " --r^2:  \t0.79\n",
      "\n",
      "The R^2 value indicates the fraction of variation inroot-to-tip\n",
      "distance explained by the sampling times.Higher values corresponds\n",
      "more clock-like behavior (max 1.0).\n",
      "\n",
      "The rate is the slope of the best fit of the date tothe root-to-tip\n",
      "distance and provides an estimate ofthe substitution rate. The rate\n",
      "needs to be positive!Negative rates suggest an inappropriate root.\n",
      "\n",
      "\n",
      "The estimated rate and tree correspond to a root date:\n",
      "\n",
      "--- root-date:\t 1924.38\n",
      "\n",
      "\n",
      "--- re-rooted tree written to \n",
      "\tprepped/div_tree/rerooted.newick\n",
      "\n",
      "--- wrote dates and root-to-tip distances to \n",
      "\tprepped/div_tree/rtt.csv\n",
      "\n",
      "--- root-to-tip plot saved to  \n",
      "\tprepped/div_tree/root_to_tip_regression.pdf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/monclalab1/anaconda3/bin/treesort\", line 5, in <module>\n",
      "    from treesort.cli import run_treesort_cli\n",
      "  File \"/Users/monclalab1/anaconda3/lib/python3.10/site-packages/treesort/cli.py\", line 15, in <module>\n",
      "    from treesort.reassortment_inference import REA_FIELD, ReassortmentDetector\n",
      "  File \"/Users/monclalab1/anaconda3/lib/python3.10/site-packages/treesort/reassortment_inference.py\", line 9, in <module>\n",
      "    from treesort.jc_reassortment_test import JCReassortmentTester, jc_pvalue\n",
      "  File \"/Users/monclalab1/anaconda3/lib/python3.10/site-packages/treesort/jc_reassortment_test.py\", line 2, in <module>\n",
      "    from scipy.stats import binomtest\n",
      "  File \"/Users/monclalab1/anaconda3/lib/python3.10/site-packages/scipy/stats/__init__.py\", line 485, in <module>\n",
      "    from ._stats_py import *\n",
      "  File \"/Users/monclalab1/anaconda3/lib/python3.10/site-packages/scipy/stats/_stats_py.py\", line 40, in <module>\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m list_of_genes \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mha\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpb2\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpb1\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mna\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnp\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpa\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmp\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m----> 2\u001b[0m \u001b[43mwrap_prep\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlist_of_genes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m          \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpreprepped\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43maln\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmetadata\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtrees\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m          \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mprepped\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mha\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmincut\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[4], line 35\u001b[0m, in \u001b[0;36mwrap_prep\u001b[0;34m(list_of_genes, home_folder, align_path, metadata_path, newick_path, output_dir, ref_gene, reroot, treesort, mincut)\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     30\u001b[0m     treesort_cmd \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     31\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtreesort\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     32\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-i\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdescriptor.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     33\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-o\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtrees_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/annotated.tree\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     34\u001b[0m             ]\n\u001b[0;32m---> 35\u001b[0m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtreesort_cmd\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/subprocess.py:505\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    503\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Popen(\u001b[38;5;241m*\u001b[39mpopenargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mas\u001b[39;00m process:\n\u001b[1;32m    504\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 505\u001b[0m         stdout, stderr \u001b[38;5;241m=\u001b[39m \u001b[43mprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcommunicate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    506\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m TimeoutExpired \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    507\u001b[0m         process\u001b[38;5;241m.\u001b[39mkill()\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/subprocess.py:1146\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1144\u001b[0m         stderr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m   1145\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m-> 1146\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1147\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1148\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/subprocess.py:1209\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1207\u001b[0m     endtime \u001b[38;5;241m=\u001b[39m _time() \u001b[38;5;241m+\u001b[39m timeout\n\u001b[1;32m   1208\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1209\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1210\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1211\u001b[0m     \u001b[38;5;66;03m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1212\u001b[0m     \u001b[38;5;66;03m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1213\u001b[0m     \u001b[38;5;66;03m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1214\u001b[0m     \u001b[38;5;66;03m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1215\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/subprocess.py:1943\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1941\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1942\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m  \u001b[38;5;66;03m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1943\u001b[0m (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_try_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1944\u001b[0m \u001b[38;5;66;03m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1945\u001b[0m \u001b[38;5;66;03m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1946\u001b[0m \u001b[38;5;66;03m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1947\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m pid \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/subprocess.py:1901\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1899\u001b[0m \u001b[38;5;124;03m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1900\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1901\u001b[0m     (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwaitpid\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwait_flags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1902\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1903\u001b[0m     \u001b[38;5;66;03m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1904\u001b[0m     \u001b[38;5;66;03m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1905\u001b[0m     \u001b[38;5;66;03m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1906\u001b[0m     pid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "list_of_genes = [\"ha\", \"pb2\",\"pb1\",\"na\",\"np\",\"pa\",\"mp\"]\n",
    "wrap_prep(list_of_genes, \n",
    "          \"preprepped\", \"aln\", \"metadata\", \"trees\", \n",
    "          \"prepped\", \"ha\", mincut =False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1276ff42-152e-43d2-bea0-7f005e35cd5c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
