{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c12d6c64-2da0-475f-ad51-c1ba00924a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import subprocess\n",
    "from datetime import datetime\n",
    "import json\n",
    "import baltic as bt\n",
    "import caffeine\n",
    "\n",
    "module_dir = \"/Users/monclalab1/Documents/scripts/\"\n",
    "sys.path.append(module_dir)\n",
    "\n",
    "from fasta_editing import fasta_to_df, fasta_writer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3b5f55c5-c69f-4999-851d-ae075ca5ee52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#treetime needs a dates.csv file to root the tree\n",
    "#the date column needs to be in decimal format\n",
    "\n",
    "def create_date_csv(strain_date_map, output_dir):\n",
    "    \n",
    "    file_name = f\"{output_dir}/strain_dates.csv\"\n",
    "    \n",
    "    with open(file_name, 'w') as date_file:\n",
    "        date_file.write(\"name, date\\n\")\n",
    "        \n",
    "        for strain, date_str in strain_date_map.items():\n",
    "\n",
    "            date = datetime.strptime(date_str, '%Y-%m-%d')\n",
    "            \n",
    "            # convert to decimal year because thats what treetime uses\n",
    "            dec_date = date.year + ((date.month - 1) * 30 + date.day) / 365.0\n",
    "        \n",
    "            date_file.write(f\"{strain}, {dec_date:.2f}\\n\")\n",
    "    \n",
    "    return file_name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7b74a4cb-c7da-4734-92f1-da22906c291c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def treesort_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "                  newick_path, output_dir, ref_gene, reroot=False):\n",
    "    \n",
    "    descriptor_entries = [] \n",
    "\n",
    "    for gene in list_of_genes:\n",
    "\n",
    "        align_file = f\"{home_folder}/{align_path}/pruned_alignments_h3nx_{gene}.fasta\"\n",
    "        metadata_file = f\"{home_folder}/{metadata_path}/metadata_h3nx_{gene}.tsv\"\n",
    "        nwk_file = f\"{home_folder}/{newick_path}/h3nx_{gene}.nwk\"\n",
    "        output_align = f\"{output_dir}/alignments/h3nx_{gene}.fasta\"\n",
    "        output_metadata = f\"{output_dir}/metadata/h3nx_{gene}.csv\"\n",
    "        output_nwk = f\"{output_dir}/div_tree/h3nx_{gene}.nwk\"\n",
    "        \n",
    "        for path in [f\"{output_dir}/alignments\", f\"{output_dir}/metadata\", f\"{output_dir}/div_tree\"]:\n",
    "            if not os.path.exists(path):\n",
    "                os.makedirs(path)\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "        # metadata file\n",
    "\n",
    "        metadata = pd.read_csv(metadata_file, sep='\\t')\n",
    "        metadata[\"date\"] = metadata[\"date\"].str.replace('XX', '01')\n",
    "        \n",
    "        # alignment file \n",
    "        align = fasta_to_df(align_file)\n",
    "        align.header = align.header.str.replace(\">\", \"\")\n",
    "\n",
    "        merged = pd.merge(align, metadata[['strain', 'date']], left_on='header', right_on='strain', how='left')\n",
    "        merged[\"header\"] = merged[['strain', 'date']].apply('|'.join, axis=1)\n",
    "        merged[\"header\"] = \">\" + merged[\"header\"]\n",
    "        fasta_writer(f\"{output_dir}/alignments/\", f\"h3nx_{gene}.fasta\", merged)\n",
    "       \n",
    "        metadata[\"strain\"] = metadata[['strain', 'date']].apply('|'.join, axis=1)\n",
    "        metadata.to_csv(output_metadata, index=False)\n",
    "\n",
    "        # stain date map for treetime\n",
    "        strain_date_map = dict(zip(merged['strain'].str.replace(\">\", \"\"), merged['date']))\n",
    "        # strain|date date map for the tree file\n",
    "        aln_date_map = dict(zip(merged['header'].str.replace(\">\", \"\"), merged['date']))\n",
    "\n",
    "        # tree file\n",
    "        \n",
    "        def replace_strain(match):\n",
    "            strain = match.group(0)[:-1]  # remove trailing ':'\n",
    "            if strain in strain_date_map:\n",
    "                return f\"{strain}|{strain_date_map[strain]}:\"\n",
    "            return match.group(0)\n",
    "        \n",
    "        with open(nwk_file, 'r') as file:\n",
    "            nwk_data = file.read()\n",
    "            \n",
    "        strain_pattern = re.compile(r'\\bA[^:]+:')\n",
    "        updated_nwk_data = strain_pattern.sub(replace_strain, nwk_data)\n",
    "        \n",
    "        with open(output_nwk, 'w') as file:\n",
    "            file.write(updated_nwk_data)\n",
    "\n",
    "\n",
    "        gene_label = f\"*{gene.upper()}\" if gene == ref_gene else gene.upper()\n",
    "\n",
    "        # rerooting using treetime if needed \n",
    "        if gene == ref_gene and reroot:\n",
    "            create_date_csv(aln_date_map, output_dir)\n",
    "            subprocess.run([\n",
    "                \"treetime\", \"clock\",\n",
    "                \"--tree\", output_nwk,\n",
    "                \"--dates\", f\"{output_dir}/strain_dates.csv\",\n",
    "                \"--aln\", output_align,\n",
    "                \"--outdir\", f\"{output_dir}/div_tree\"\n",
    "            ], check=True)\n",
    "            descriptor_entries.append([gene_label, output_align, f\"{output_dir}/div_tree/rerooted.newick\"])\n",
    "        else:\n",
    "            descriptor_entries.append([gene_label, output_align, output_nwk])\n",
    "\n",
    "    # descriptor csv\n",
    "    with open(\"descriptor.csv\", 'w') as descriptor_file:\n",
    "        for row in descriptor_entries:\n",
    "            descriptor_file.write(','.join(row) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bd4d9cec-ef5b-401f-9a51-907b27ea5db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def wrap_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "              newick_path, output_dir, ref_gene, \n",
    "              reroot=False, treesort=True, mincut=False):\n",
    "    \n",
    "    \n",
    "    treesort_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "                  newick_path, output_dir, ref_gene, reroot = False)\n",
    "    \n",
    "    if treesort:\n",
    "        \n",
    "        if not os.path.exists(\"trees\"):\n",
    "            os.makedirs(\"trees\")\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "        caffeine.on(display=False)\n",
    "        \n",
    "        if mincut:\n",
    "            \n",
    "            treesort_cmd = [\n",
    "                        \"treesort\",\n",
    "                        \"-i\", \"descriptor.csv\",\n",
    "                        \"-o\", \"trees/annotated.tree\"\n",
    "                        \"-m\", \"mincut\"\n",
    "                    ]\n",
    "        else:\n",
    "        \n",
    "            treesort_cmd = [\n",
    "                        \"treesort\",\n",
    "                        \"-i\", \"descriptor.csv\",\n",
    "                        \"-o\", \"trees/annotated.tree\"\n",
    "                    ]\n",
    "        subprocess.run(treesort_cmd, check=True, capture_output=True, text=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c1a42033-4fc4-4ea9-9356-533612d8d68b",
   "metadata": {},
   "outputs": [
    {
     "ename": "CalledProcessError",
     "evalue": "Command '['treesort', '-i', 'descriptor.csv', '-o', 'trees/annotated.tree-m', 'mincut']' returned non-zero exit status 2.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m list_of_genes \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mha\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpb2\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpb1\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mna\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnp\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpa\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mns\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmp\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m----> 2\u001b[0m \u001b[43mwrap_prep\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlist_of_genes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m          \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpreprepped\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43maln\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmetadata\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtrees\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m          \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mprepped\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mha\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmincut\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[12], line 33\u001b[0m, in \u001b[0;36mwrap_prep\u001b[0;34m(list_of_genes, home_folder, align_path, metadata_path, newick_path, output_dir, ref_gene, reroot, treesort, mincut)\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     28\u001b[0m     treesort_cmd \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m     29\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtreesort\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     30\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-i\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdescriptor.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     31\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-o\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtrees/annotated.tree\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     32\u001b[0m             ]\n\u001b[0;32m---> 33\u001b[0m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtreesort_cmd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcapture_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/subprocess.py:526\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    524\u001b[0m     retcode \u001b[38;5;241m=\u001b[39m process\u001b[38;5;241m.\u001b[39mpoll()\n\u001b[1;32m    525\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m check \u001b[38;5;129;01mand\u001b[39;00m retcode:\n\u001b[0;32m--> 526\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(retcode, process\u001b[38;5;241m.\u001b[39margs,\n\u001b[1;32m    527\u001b[0m                                  output\u001b[38;5;241m=\u001b[39mstdout, stderr\u001b[38;5;241m=\u001b[39mstderr)\n\u001b[1;32m    528\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m CompletedProcess(process\u001b[38;5;241m.\u001b[39margs, retcode, stdout, stderr)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command '['treesort', '-i', 'descriptor.csv', '-o', 'trees/annotated.tree-m', 'mincut']' returned non-zero exit status 2."
     ]
    }
   ],
   "source": [
    "list_of_genes = [\"ha\", \"pb2\",\"pb1\",\"na\",\"np\",\"pa\",\"ns\",\"mp\"]\n",
    "wrap_prep(list_of_genes, \n",
    "          \"preprepped\", \"aln\", \"metadata\", \"trees\", \n",
    "          \"prepped\", \"ha\", mincut =True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1276ff42-152e-43d2-bea0-7f005e35cd5c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
