{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "aa0bd7bf-234a-4aa2-a020-e0e4c1b55074",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import baltic as bt\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "from collections import Counter\n",
    "\n",
    "module_dir = \"/Users/monclalab1/Documents/scripts/\"\n",
    "sys.path.append(module_dir)\n",
    "\n",
    "from fasta_editing import fasta_to_df, fasta_writer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a1845767-0a9c-45fd-88e6-9e812d2cebed",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = fasta_to_df(\"utr_h3nx_na.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1222c4e4-ec8b-4bea-8e25-cf9dbb6d863f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_values(['gctttcgcaattttagaaaaaactccttgt-ttctact-', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gcttcgcaaattttag-----------------------', 'gcttcgcaaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gcttcgcaaattttag-----------------------', 'gcttcgcaatttttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaact--------------', 'gctttcgcaattttagaaaaaaact--------------', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttaaaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gcttcgcaaattttag-----------------------', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattc--------------------------', 'gctttcgcaattttagaaaaaactccttgttt--ctact', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctg-----------------------------------', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaactccttgttt--ctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaa-----------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gcttcgcaaattttag-----------------------', 'gcttcgcaaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gcttcgcaaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttagaaaaaaact--------------', 'gctttcgcaattttagaaaaaaact--------------', 'gctttcgcaattttagaaaaaaact--------------', 'gctttcgcaattttagaaaaaaact--------------', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaactccttgt-ttctact-', 'gctttcgcaattttag-----------------------', 'gcttcgcaaattttag-----------------------', 'gcttcgcaaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gcttcgcaaattttag-----------------------', 'gctttctcaattttag-----------------------', 'gctttcgcaattttagaaaaaaactccttg-tttc----', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaactccttgttt--ctact', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaactccttgt-ttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctayg-', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaactccttgt-ttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttgcgcc-', 'gctttcgcaattttagaaaaaaactccttgtttctcgc-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gctttcgcaattttagaaaaaaactccttgtttctacc-', 'gctttcgcaattttagaaaaaaactccttgtttctact-', 'gcttcgcaaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttag-----------------------', 'gctttcgcaattttagaaaaaactccttgt-ttctact-', 'gctttcgcaattttagaaaaaactccttgt-ttctact-', 'gctttcgcaattttagaaaaaaactccttg-tttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaaactccttgtttctact', 'gctttcgcaattttagaaaaaaactccttg-tttctact'])\n"
     ]
    }
   ],
   "source": [
    "df['strain'] = df['header'].str.split(\"|\").str[0].str.extract(r'>(.*)')\n",
    "df['sequence'] = df['sequence'].str.lower()\n",
    "\n",
    "#starting UTR\n",
    "# the sequences that have the UTR will not start with atg or a gap\n",
    "df_with_start_utr = df[~df['sequence'].str.startswith(('atg', '-'))].copy()\n",
    "df_with_start_utr['start_UTR'] = df_with_start_utr['sequence'].str.split(\"atg\", n=1).str[0]\n",
    "start_utr_dict = pd.Series(df_with_start_utr['start_UTR'].values, index=df_with_start_utr['strain']).to_dict()\n",
    "\n",
    "# print(start_utr_dict.values())\n",
    "\n",
    "\n",
    "#ending UTR\n",
    "\n",
    "def extract_end_utr_by_position(sequence):\n",
    "    if len(sequence) > 1448:  # Ensure sequence extends beyond position 1466 (1-based index)\n",
    "        utr_sequence = sequence[1448:]  # Extract sequence after position 1466\n",
    "        return utr_sequence if utr_sequence.strip(\"-\") else None  # Ignore gap-only UTRs\n",
    "    return None  # Return None if sequence isn't long enough\n",
    "\n",
    "df[\"end_utr\"] = df[\"sequence\"].apply(extract_end_utr_by_position)\n",
    "end_utr_dict = df.dropna(subset=[\"end_utr\"]).set_index(\"strain\")[\"end_utr\"].to_dict()\n",
    "\n",
    "print(end_utr_dict.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "90a3ad20-4bd8-42b7-8bc9-5f067dc683bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Tree height: 17.931000\n",
      "Tree length: 199.609000\n",
      "annotations present\n",
      "\n",
      "Numbers of objects in tree: 524 (230 nodes and 294 leaves)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mytree, mymeta = bt.loadJSON(\"canine_build/auspice/canine_h3nx_na.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6bdc445c-1c39-49e6-a85f-33d5ffa41885",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NODE_0000102 is the branch of interest with 4 mutations \n",
    "# T16A, Y67H, H155Y, R283Q\n",
    "\n",
    "target_node= \"NODE_0000102\"\n",
    "\n",
    "# all strains descending from branch\n",
    "for k in mytree.Objects:\n",
    "    \n",
    "    if k.name == target_node:\n",
    "        \n",
    "        subtree = mytree.subtree(k)\n",
    "        \n",
    "        descending_leaves = {leaf.name for leaf in subtree.getExternal()}\n",
    "\n",
    "# all strains preceding branch \n",
    "\n",
    "preceding_leaves = set()\n",
    "\n",
    "for j in mytree.Objects:\n",
    "    \n",
    "    if j.name in descending_leaves:\n",
    "        continue\n",
    "        \n",
    "    elif j.is_leaf():\n",
    "        preceding_leaves.add(j.name)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2146b5cf-b3e9-416c-a13e-b78c7323ef9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Consensus starting UTR before branch:\n",
      "Counter({'a': 53})\n",
      "Counter({'g': 53})\n",
      "Counter({'c': 53})\n",
      "Counter({'a': 50, 'g': 3})\n",
      "Counter({'a': 53})\n",
      "Counter({'a': 53})\n",
      "Counter({'a': 53})\n",
      "Counter({'g': 53})\n",
      "Counter({'c': 53})\n",
      "Counter({'a': 53})\n",
      "Counter({'g': 53})\n",
      "Counter({'g': 53})\n",
      "Counter({'a': 53})\n",
      "Counter({'g': 53})\n",
      "Counter({'t': 51, 'g': 2})\n",
      "Counter({'a': 50, 'g': 3})\n",
      "Counter({'a': 53})\n",
      "Counter({'a': 51, 'g': 2})\n",
      "Counter({'a': 52, 'g': 1})\n",
      "agcaaaagcaggagtaaaa\n",
      " \n",
      "Consensus starting UTR after branch:\n",
      "Counter({'a': 65})\n",
      "Counter({'g': 65})\n",
      "Counter({'c': 65})\n",
      "Counter({'a': 56, 'n': 6, 'r': 2, 'g': 1})\n",
      "Counter({'a': 65})\n",
      "Counter({'a': 65})\n",
      "Counter({'a': 65})\n",
      "Counter({'g': 65})\n",
      "Counter({'c': 65})\n",
      "Counter({'a': 65})\n",
      "Counter({'g': 65})\n",
      "Counter({'g': 65})\n",
      "Counter({'a': 65})\n",
      "Counter({'g': 65})\n",
      "Counter({'t': 65})\n",
      "Counter({'a': 65})\n",
      "Counter({'a': 65})\n",
      "Counter({'a': 65})\n",
      "Counter({'a': 65})\n",
      "agcaaaagcaggagtaaaa\n"
     ]
    }
   ],
   "source": [
    "# consensus starting UTR for all descending and preceding leaves\n",
    "descending_start_utr_seq = [start_utr_dict[strain] for strain in descending_leaves if strain in start_utr_dict]\n",
    "preceding_start_utr_seq = [start_utr_dict[strain] for strain in preceding_leaves if strain in start_utr_dict]\n",
    "\n",
    "def get_consensus(utr_list):\n",
    "    \"\"\"Computes the consensus sequence from a list of aligned UTR sequences.\"\"\"\n",
    "    if not utr_list:\n",
    "        return None\n",
    "\n",
    "    consensus_sequence = []\n",
    "\n",
    "    for position in zip(*utr_list):  \n",
    "        nucleotide_counts = Counter(position)\n",
    "        print(nucleotide_counts)\n",
    "        most_common_nucleotide = nucleotide_counts.most_common(1)[0][0]\n",
    "        consensus_sequence.append(most_common_nucleotide) \n",
    "\n",
    "    return \"\".join(consensus_sequence)\n",
    "\n",
    "print(\"Consensus starting UTR before branch:\")\n",
    "start_consensus_pre_utr = get_consensus(preceding_start_utr_seq)\n",
    "print(start_consensus_pre_utr)\n",
    "\n",
    "print(\" \")\n",
    "\n",
    "print(\"Consensus starting UTR after branch:\")\n",
    "start_consensus_des_utr = get_consensus(descending_start_utr_seq)\n",
    "print(start_consensus_des_utr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "830c0850-1b4f-4708-a683-206312465dad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Consensus ending UTR before branch:\n",
      "Counter({'g': 54})\n",
      "Counter({'c': 54})\n",
      "Counter({'t': 54})\n",
      "Counter({'t': 54})\n",
      "Counter({'t': 54})\n",
      "Counter({'c': 54})\n",
      "Counter({'g': 54})\n",
      "Counter({'c': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'t': 54})\n",
      "Counter({'t': 54})\n",
      "Counter({'t': 54})\n",
      "Counter({'t': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'g': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'a': 54})\n",
      "Counter({'a': 51, 'c': 2, '-': 1})\n",
      "Counter({'c': 47, 'a': 4, 't': 2, '-': 1})\n",
      "Counter({'t': 47, 'c': 6, '-': 1})\n",
      "Counter({'c': 49, 't': 4, '-': 1})\n",
      "Counter({'c': 51, 't': 2, '-': 1})\n",
      "Counter({'t': 49, 'c': 4, '-': 1})\n",
      "Counter({'t': 51, 'g': 2, '-': 1})\n",
      "Counter({'g': 47, 't': 6, '-': 1})\n",
      "Counter({'-': 47, 'g': 4, 't': 3})\n",
      "Counter({'t': 53, '-': 1})\n",
      "Counter({'t': 51, '-': 3})\n",
      "Counter({'t': 50, '-': 3, 'c': 1})\n",
      "Counter({'c': 52, 't': 1, '-': 1})\n",
      "Counter({'t': 52, 'a': 1, '-': 1})\n",
      "Counter({'a': 52, 'c': 1, '-': 1})\n",
      "Counter({'c': 52, 't': 1, '-': 1})\n",
      "Counter({'t': 52, '-': 2})\n",
      "gctttcgcaattttagaaaaaaactccttg-tttctact\n",
      " \n",
      "Consensus ending UTR after branch:\n",
      "Counter({'g': 124})\n",
      "Counter({'c': 124})\n",
      "Counter({'t': 124})\n",
      "Counter({'t': 123, 'g': 1})\n",
      "Counter({'t': 116, 'c': 7, '-': 1})\n",
      "Counter({'c': 116, 'g': 7, '-': 1})\n",
      "Counter({'g': 116, 'c': 7, '-': 1})\n",
      "Counter({'c': 116, 'a': 7, '-': 1})\n",
      "Counter({'a': 123, '-': 1})\n",
      "Counter({'a': 123, '-': 1})\n",
      "Counter({'t': 123, '-': 1})\n",
      "Counter({'t': 123, '-': 1})\n",
      "Counter({'t': 122, 'c': 1, '-': 1})\n",
      "Counter({'t': 122, '-': 2})\n",
      "Counter({'a': 122, '-': 2})\n",
      "Counter({'g': 122, '-': 2})\n",
      "Counter({'a': 72, '-': 52})\n",
      "Counter({'a': 72, '-': 52})\n",
      "Counter({'a': 72, '-': 52})\n",
      "Counter({'a': 72, '-': 52})\n",
      "Counter({'a': 72, '-': 52})\n",
      "Counter({'a': 72, '-': 52})\n",
      "Counter({'a': 67, '-': 52, 'c': 5})\n",
      "Counter({'c': 56, '-': 52, 'a': 11, 't': 5})\n",
      "Counter({'t': 56, '-': 52, 'c': 16})\n",
      "Counter({'-': 58, 'c': 55, 't': 11})\n",
      "Counter({'c': 61, '-': 58, 't': 5})\n",
      "Counter({'-': 58, 't': 55, 'c': 11})\n",
      "Counter({'t': 61, '-': 58, 'g': 5})\n",
      "Counter({'-': 58, 'g': 50, 't': 16})\n",
      "Counter({'-': 85, 't': 28, 'g': 11})\n",
      "Counter({'t': 66, '-': 58})\n",
      "Counter({'t': 66, '-': 58})\n",
      "Counter({'-': 58, 't': 33, 'c': 32, 'g': 1})\n",
      "Counter({'-': 58, 'c': 34, 't': 32})\n",
      "Counter({'-': 59, 't': 32, 'a': 31, 'c': 1, 'g': 1})\n",
      "Counter({'-': 59, 'a': 32, 'c': 31, 'g': 1, 'y': 1})\n",
      "Counter({'-': 59, 'c': 35, 't': 29, 'g': 1})\n",
      "Counter({'-': 92, 't': 32})\n",
      "gctttcgcaattttagaaaaaaact-c-t--tt------\n"
     ]
    }
   ],
   "source": [
    "# consensus starting UTR for all descending and preceding leaves\n",
    "descending_end_utr_seq = [end_utr_dict[strain] for strain in descending_leaves if strain in end_utr_dict]\n",
    "preceding_end_utr_seq = [end_utr_dict[strain] for strain in preceding_leaves if strain in end_utr_dict]\n",
    "\n",
    "def get_consensus(utr_list):\n",
    "    \"\"\"Computes the consensus sequence from a list of aligned UTR sequences.\"\"\"\n",
    "    if not utr_list:\n",
    "        return None\n",
    "\n",
    "    consensus_sequence = []\n",
    "\n",
    "    for position in zip(*utr_list):  \n",
    "        nucleotide_counts = Counter(position)\n",
    "        print(nucleotide_counts)\n",
    "        most_common_nucleotide = nucleotide_counts.most_common(1)[0][0]\n",
    "        consensus_sequence.append(most_common_nucleotide) \n",
    "\n",
    "    return \"\".join(consensus_sequence)\n",
    "\n",
    "print(\"Consensus ending UTR before branch:\")\n",
    "end_consensus_pre_utr = get_consensus(preceding_end_utr_seq)\n",
    "print(end_consensus_pre_utr)\n",
    "\n",
    "print(\" \")\n",
    "\n",
    "print(\"Consensus ending UTR after branch:\")\n",
    "end_consensus_des_utr = get_consensus(descending_end_utr_seq)\n",
    "print(end_consensus_des_utr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5fe7cfec-9997-4480-beea-07e8b4a2e47c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'a': 134})\n",
      "Counter({'g': 134})\n",
      "Counter({'c': 134})\n",
      "Counter({'a': 121, 'n': 7, 'g': 4, 'r': 2})\n",
      "Counter({'a': 134})\n",
      "Counter({'a': 134})\n",
      "Counter({'a': 134})\n",
      "Counter({'g': 134})\n",
      "Counter({'c': 134})\n",
      "Counter({'a': 134})\n",
      "Counter({'g': 134})\n",
      "Counter({'g': 134})\n",
      "Counter({'a': 134})\n",
      "Counter({'g': 134})\n",
      "Counter({'t': 132, 'g': 2})\n",
      "Counter({'a': 130, 'g': 4})\n",
      "Counter({'a': 134})\n",
      "Counter({'a': 132, 'g': 2})\n",
      "Counter({'a': 132, 'g': 2})\n",
      "Consensus UTR across tree: agcaaaagcaggagtaaaa\n"
     ]
    }
   ],
   "source": [
    "#all starting UTRs\n",
    "\n",
    "uni_start_utr_sequences = [start_utr_dict[strain] for strain in start_utr_dict]\n",
    "\n",
    "def get_consensus(utr_list):\n",
    "    \"\"\"Computes the consensus sequence from a list of aligned UTR sequences.\"\"\"\n",
    "    if not utr_list:\n",
    "        return None\n",
    "\n",
    "    consensus_sequence = []\n",
    "\n",
    "    for position in zip(*utr_list):  \n",
    "        nucleotide_counts = Counter(position)\n",
    "        print(nucleotide_counts)\n",
    "        most_common_nucleotide = nucleotide_counts.most_common(1)[0][0]\n",
    "        consensus_sequence.append(most_common_nucleotide) \n",
    "\n",
    "    return \"\".join(consensus_sequence)\n",
    "\n",
    "uni_start_consensus_utr = get_consensus(uni_start_utr_sequences)\n",
    "\n",
    "print(\"Consensus UTR across tree:\", uni_start_consensus_utr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a354b78d-27ed-46bf-a2f6-14fc8ac98d76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'g': 209})\n",
      "Counter({'c': 209})\n",
      "Counter({'t': 209})\n",
      "Counter({'t': 208, 'g': 1})\n",
      "Counter({'t': 196, 'c': 12, '-': 1})\n",
      "Counter({'c': 196, 'g': 12, '-': 1})\n",
      "Counter({'g': 195, 'c': 12, '-': 1, 't': 1})\n",
      "Counter({'c': 196, 'a': 12, '-': 1})\n",
      "Counter({'a': 208, '-': 1})\n",
      "Counter({'a': 207, 't': 1, '-': 1})\n",
      "Counter({'t': 208, '-': 1})\n",
      "Counter({'t': 208, '-': 1})\n",
      "Counter({'t': 207, 'c': 1, '-': 1})\n",
      "Counter({'t': 207, '-': 2})\n",
      "Counter({'a': 207, '-': 2})\n",
      "Counter({'g': 206, '-': 2, 'a': 1})\n",
      "Counter({'a': 142, '-': 67})\n",
      "Counter({'a': 142, '-': 67})\n",
      "Counter({'a': 142, '-': 67})\n",
      "Counter({'a': 142, '-': 67})\n",
      "Counter({'a': 142, '-': 67})\n",
      "Counter({'a': 142, '-': 67})\n",
      "Counter({'a': 132, '-': 68, 'c': 9})\n",
      "Counter({'c': 113, '-': 68, 'a': 19, 't': 9})\n",
      "Counter({'t': 113, '-': 68, 'c': 28})\n",
      "Counter({'c': 116, '-': 74, 't': 19})\n",
      "Counter({'c': 126, '-': 74, 't': 9})\n",
      "Counter({'t': 116, '-': 74, 'c': 19})\n",
      "Counter({'t': 126, '-': 74, 'g': 9})\n",
      "Counter({'g': 107, '-': 74, 't': 28})\n",
      "Counter({'-': 158, 't': 32, 'g': 19})\n",
      "Counter({'t': 135, '-': 74})\n",
      "Counter({'t': 132, '-': 77})\n",
      "Counter({'t': 97, '-': 77, 'c': 34, 'g': 1})\n",
      "Counter({'c': 101, '-': 74, 't': 34})\n",
      "Counter({'t': 99, '-': 75, 'a': 33, 'g': 1, 'c': 1})\n",
      "Counter({'a': 99, '-': 75, 'c': 33, 'y': 1, 'g': 1})\n",
      "Counter({'c': 102, '-': 75, 't': 31, 'g': 1})\n",
      "Counter({'-': 110, 't': 99})\n",
      "Consensus ending UTR across tree: gctttcgcaattttagaaaaaaactccttg-tttctac-\n"
     ]
    }
   ],
   "source": [
    "#all ending UTRs\n",
    "\n",
    "uni_end_utr_sequences = [end_utr_dict[strain] for strain in end_utr_dict]\n",
    "\n",
    "def get_consensus(utr_list):\n",
    "    \"\"\"Computes the consensus sequence from a list of aligned UTR sequences.\"\"\"\n",
    "    if not utr_list:\n",
    "        return None\n",
    "\n",
    "    consensus_sequence = []\n",
    "\n",
    "    for position in zip(*utr_list):  \n",
    "        nucleotide_counts = Counter(position)\n",
    "        print(nucleotide_counts)\n",
    "        most_common_nucleotide = nucleotide_counts.most_common(1)[0][0]\n",
    "        consensus_sequence.append(most_common_nucleotide) \n",
    "\n",
    "    return \"\".join(consensus_sequence)\n",
    "\n",
    "uni_end_consensus_utr = get_consensus(uni_end_utr_sequences)\n",
    "\n",
    "print(\"Consensus ending UTR across tree:\", uni_end_consensus_utr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1dbe83b-beee-4c14-a330-6f4ae18f7d77",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46e6fdb6-6a85-4789-9cba-a36d7eae345d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65b4a003-25ae-4053-b5ea-457919ff6db3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
