{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c12d6c64-2da0-475f-ad51-c1ba00924a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import subprocess\n",
    "from datetime import datetime\n",
    "import json\n",
    "import baltic as bt\n",
    "import caffeine\n",
    "\n",
    "module_dir = \"/Users/monclalab1/Documents/scripts/\"\n",
    "sys.path.append(module_dir)\n",
    "\n",
    "from fasta_editing import fasta_to_df, fasta_writer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3b5f55c5-c69f-4999-851d-ae075ca5ee52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#treetime needs a dates.csv file to root the tree\n",
    "#the date column needs to be in decimal format\n",
    "\n",
    "def create_date_csv(strain_date_map, output_dir):\n",
    "    \n",
    "    file_name = f\"{output_dir}/strain_dates.csv\"\n",
    "    \n",
    "    with open(file_name, 'w') as date_file:\n",
    "        date_file.write(\"name, date\\n\")\n",
    "        \n",
    "        for strain, date_str in strain_date_map.items():\n",
    "\n",
    "            date = datetime.strptime(date_str, '%Y-%m-%d')\n",
    "            \n",
    "            # convert to decimal year because thats what treetime uses\n",
    "            dec_date = date.year + ((date.month - 1) * 30 + date.day) / 365.0\n",
    "        \n",
    "            date_file.write(f\"{strain}, {dec_date:.2f}\\n\")\n",
    "    \n",
    "    return file_name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7b74a4cb-c7da-4734-92f1-da22906c291c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def treesort_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "                  newick_path, output_dir, ref_gene, reroot=False):\n",
    "    \n",
    "    descriptor_entries = [] \n",
    "\n",
    "    for gene in list_of_genes:\n",
    "\n",
    "        align_file = f\"{home_folder}/{align_path}/h3nx_{gene}.fasta\"\n",
    "        metadata_file = f\"{home_folder}/{metadata_path}/metadata_h3nx_{gene}.tsv\"\n",
    "        nwk_file = f\"{home_folder}/{newick_path}/h3nx_{gene}.nwk\"\n",
    "        output_align = f\"{output_dir}/alignments/h3nx_{gene}.fasta\"\n",
    "        output_metadata = f\"{output_dir}/metadata/h3nx_{gene}.csv\"\n",
    "        output_nwk = f\"{output_dir}/div_tree/h3nx_{gene}.nwk\"\n",
    "        \n",
    "        for path in [f\"{output_dir}/alignments\", f\"{output_dir}/metadata\", f\"{output_dir}/div_tree\"]:\n",
    "            if not os.path.exists(path):\n",
    "                os.makedirs(path)\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "        # metadata file\n",
    "\n",
    "        metadata = pd.read_csv(metadata_file, sep='\\t')\n",
    "        metadata[\"date\"] = metadata[\"date\"].str.replace('XX', '01')\n",
    "        \n",
    "        # alignment file \n",
    "        align = fasta_to_df(align_file)\n",
    "        align.header = align.header.str.replace(\">\", \"\")\n",
    "\n",
    "        merged = pd.merge(align, metadata[['strain', 'date']], left_on='header', right_on='strain', how='left')\n",
    "        merged[\"header\"] = merged[['strain', 'date']].apply('|'.join, axis=1)\n",
    "        merged[\"header\"] = \">\" + merged[\"header\"]\n",
    "        fasta_writer(f\"{output_dir}/alignments/\", f\"h3nx_{gene}.fasta\", merged)\n",
    "       \n",
    "        metadata[\"strain\"] = metadata[['strain', 'date']].apply('|'.join, axis=1)\n",
    "        metadata.to_csv(output_metadata, index=False)\n",
    "\n",
    "        # stain date map for treetime\n",
    "        strain_date_map = dict(zip(merged['strain'].str.replace(\">\", \"\"), merged['date']))\n",
    "        # strain|date date map for the tree file\n",
    "        aln_date_map = dict(zip(merged['header'].str.replace(\">\", \"\"), merged['date']))\n",
    "\n",
    "        # tree file\n",
    "        \n",
    "        def replace_strain(match):\n",
    "            strain = match.group(0)[:-1]  # remove trailing ':'\n",
    "            if strain in strain_date_map:\n",
    "                return f\"{strain}|{strain_date_map[strain]}:\"\n",
    "            return match.group(0)\n",
    "        \n",
    "        with open(nwk_file, 'r') as file:\n",
    "            nwk_data = file.read()\n",
    "            \n",
    "        strain_pattern = re.compile(r'\\bA[^:]+:')\n",
    "        updated_nwk_data = strain_pattern.sub(replace_strain, nwk_data)\n",
    "        \n",
    "        with open(output_nwk, 'w') as file:\n",
    "            file.write(updated_nwk_data)\n",
    "\n",
    "\n",
    "        gene_label = f\"*{gene.upper()}\" if gene == ref_gene else gene.upper()\n",
    "\n",
    "        # rerooting using treetime if needed \n",
    "        if gene == ref_gene and reroot:\n",
    "            create_date_csv(aln_date_map, output_dir)\n",
    "            subprocess.run([\n",
    "                \"treetime\", \"clock\",\n",
    "                \"--tree\", output_nwk,\n",
    "                \"--dates\", f\"{output_dir}/strain_dates.csv\",\n",
    "                \"--aln\", output_align,\n",
    "                \"--outdir\", f\"{output_dir}/div_tree\"\n",
    "            ], check=True)\n",
    "            descriptor_entries.append([gene_label, output_align, f\"{output_dir}/div_tree/rerooted.newick\"])\n",
    "        else:\n",
    "            descriptor_entries.append([gene_label, output_align, output_nwk])\n",
    "\n",
    "    # descriptor csv\n",
    "    with open(\"descriptor.csv\", 'w') as descriptor_file:\n",
    "        for row in descriptor_entries:\n",
    "            descriptor_file.write(','.join(row) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bd4d9cec-ef5b-401f-9a51-907b27ea5db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def wrap_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "              newick_path, output_dir, ref_gene, \n",
    "              reroot=True, treesort=True, mincut=False):\n",
    "    \n",
    "    \n",
    "    treesort_prep(list_of_genes, home_folder, align_path, metadata_path,\n",
    "                  newick_path, output_dir, ref_gene, reroot = True)\n",
    "    \n",
    "    if treesort:\n",
    "        \n",
    "        for i in range(1,6):\n",
    "        \n",
    "            if not os.path.exists(f\"trees_{i}\"):\n",
    "                os.makedirs(f\"trees_{i}\")\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "            caffeine.on(display=False)\n",
    "\n",
    "            if mincut:\n",
    "\n",
    "                treesort_cmd = [\n",
    "                            \"treesort\",\n",
    "                            \"-i\", \"descriptor.csv\",\n",
    "                            \"-o\", f\"trees_{i}/annotated.tree\",\n",
    "                            \"-m\", \"mincut\"\n",
    "                        ]\n",
    "            else:\n",
    "\n",
    "                treesort_cmd = [\n",
    "                            \"treesort\",\n",
    "                            \"-i\", \"descriptor.csv\",\n",
    "                            \"-o\", f\"trees_{i}/annotated.tree\"\n",
    "                        ]\n",
    "            subprocess.run(treesort_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c1a42033-4fc4-4ea9-9356-533612d8d68b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "0.00\t-TreeAnc: set-up\n",
      "\n",
      "Attempting to parse dates...\n",
      "\tUsing column 'name' as name. This needs match the taxon names in the tree!!\n",
      "\tUsing column 'date' as date.\n",
      "\n",
      "0.35\tTreeTime.reroot: with method or node: least-squares\n",
      "\n",
      "0.36\tTreeTime.reroot: rerooting will ignore covariance and shared ancestry.\n",
      "\n",
      "0.70\tTreeTime.reroot: with method or node: least-squares\n",
      "\n",
      "0.70\tTreeTime.reroot: rerooting will ignore covariance and shared ancestry.\n",
      "\n",
      "1.01\tTreeTime.reroot: with method or node: least-squares\n",
      "\n",
      "1.01\tTreeTime.reroot: rerooting will ignore covariance and shared ancestry.\n",
      "\n",
      " Root-Tip-Regression:\n",
      " --rate:\t1.318e-03\n",
      " --r^2:  \t0.79\n",
      "\n",
      "The R^2 value indicates the fraction of variation inroot-to-tip\n",
      "distance explained by the sampling times.Higher values corresponds\n",
      "more clock-like behavior (max 1.0).\n",
      "\n",
      "The rate is the slope of the best fit of the date tothe root-to-tip\n",
      "distance and provides an estimate ofthe substitution rate. The rate\n",
      "needs to be positive!Negative rates suggest an inappropriate root.\n",
      "\n",
      "\n",
      "The estimated rate and tree correspond to a root date:\n",
      "\n",
      "--- root-date:\t 1924.38\n",
      "\n",
      "\n",
      "--- re-rooted tree written to \n",
      "\tprepped/div_tree/rerooted.newick\n",
      "\n",
      "--- wrote dates and root-to-tip distances to \n",
      "\tprepped/div_tree/rtt.csv\n",
      "\n",
      "--- root-to-tip plot saved to  \n",
      "\tprepped/div_tree/root_to_tip_regression.pdf\n",
      "P-value threshold for significance was set to 0.001\n",
      "Read 7 segments: HA, PB2, PB1, NA, NP, PA, MP\n",
      "Estimating molecular clock rates for each segment (TreeTime)...\n",
      "\tHA estimated molecular clock rate: 0.0013003 (R^2 = 0.889)\n",
      "\tPB2 estimated molecular clock rate: 0.0016705 (R^2 = 0.934)\n",
      "\tPB1 estimated molecular clock rate: 0.0013672 (R^2 = 0.948)\n",
      "\tNA estimated molecular clock rate: 0.0012655 (R^2 = 0.914)\n",
      "\tNP estimated molecular clock rate: 0.0012612 (R^2 = 0.882)\n",
      "\tPA estimated molecular clock rate: 0.0014513 (R^2 = 0.917)\n",
      "\tMP estimated molecular clock rate: 0.0004999 (R^2 = 0.848)\n",
      "Found 627 strains in common across the alignments.\n",
      "Optimally resolving the multifurcations to minimize reassortment...\n",
      "Inferring reassortment with the PB2 segment...\n",
      "\tInferred reassortment events with PB2: 124.\n",
      "Inferring reassortment with the PB1 segment...\n",
      "\tInferred reassortment events with PB1: 90.\n",
      "Inferring reassortment with the NA segment...\n",
      "\tInferred reassortment events with NA: 53.\n",
      "Inferring reassortment with the NP segment...\n",
      "\tInferred reassortment events with NP: 116.\n",
      "Inferring reassortment with the PA segment...\n",
      "\tInferred reassortment events with PA: 145.\n",
      "Inferring reassortment with the MP segment...\n",
      "\tInferred reassortment events with MP: 96.\n",
      "\n",
      "Estimated reassortment rate per ancestral lineage per year.\n",
      "\tBernoulli estimate: 0.21208\n",
      "\tPoisson estimate (for dense datasets): 0.202527\n",
      "Saved the annotated tree file to trees_1/annotated.tree\n",
      "P-value threshold for significance was set to 0.001\n",
      "Read 7 segments: HA, PB2, PB1, NA, NP, PA, MP\n",
      "Estimating molecular clock rates for each segment (TreeTime)...\n",
      "\tHA estimated molecular clock rate: 0.0013003 (R^2 = 0.889)\n",
      "\tPB2 estimated molecular clock rate: 0.0016705 (R^2 = 0.934)\n",
      "\tPB1 estimated molecular clock rate: 0.0013672 (R^2 = 0.948)\n",
      "\tNA estimated molecular clock rate: 0.0012655 (R^2 = 0.914)\n",
      "\tNP estimated molecular clock rate: 0.0012612 (R^2 = 0.882)\n",
      "\tPA estimated molecular clock rate: 0.0014513 (R^2 = 0.917)\n",
      "\tMP estimated molecular clock rate: 0.0004999 (R^2 = 0.848)\n",
      "Found 627 strains in common across the alignments.\n",
      "Optimally resolving the multifurcations to minimize reassortment...\n",
      "Inferring reassortment with the PB2 segment...\n",
      "\tInferred reassortment events with PB2: 122.\n",
      "Inferring reassortment with the PB1 segment...\n",
      "\tInferred reassortment events with PB1: 86.\n",
      "Inferring reassortment with the NA segment...\n",
      "\tInferred reassortment events with NA: 53.\n",
      "Inferring reassortment with the NP segment...\n",
      "\tInferred reassortment events with NP: 118.\n",
      "Inferring reassortment with the PA segment...\n",
      "\tInferred reassortment events with PA: 142.\n",
      "Inferring reassortment with the MP segment...\n",
      "\tInferred reassortment events with MP: 95.\n",
      "\n",
      "Estimated reassortment rate per ancestral lineage per year.\n",
      "\tBernoulli estimate: 0.210351\n",
      "\tPoisson estimate (for dense datasets): 0.200344\n",
      "Saved the annotated tree file to trees_2/annotated.tree\n",
      "P-value threshold for significance was set to 0.001\n",
      "Read 7 segments: HA, PB2, PB1, NA, NP, PA, MP\n",
      "Estimating molecular clock rates for each segment (TreeTime)...\n",
      "\tHA estimated molecular clock rate: 0.0013003 (R^2 = 0.889)\n",
      "\tPB2 estimated molecular clock rate: 0.0016705 (R^2 = 0.934)\n",
      "\tPB1 estimated molecular clock rate: 0.0013672 (R^2 = 0.948)\n",
      "\tNA estimated molecular clock rate: 0.0012655 (R^2 = 0.914)\n",
      "\tNP estimated molecular clock rate: 0.0012612 (R^2 = 0.882)\n",
      "\tPA estimated molecular clock rate: 0.0014513 (R^2 = 0.917)\n",
      "\tMP estimated molecular clock rate: 0.0004999 (R^2 = 0.848)\n",
      "Found 627 strains in common across the alignments.\n",
      "Optimally resolving the multifurcations to minimize reassortment...\n",
      "Inferring reassortment with the PB2 segment...\n",
      "\tInferred reassortment events with PB2: 124.\n",
      "Inferring reassortment with the PB1 segment...\n",
      "\tInferred reassortment events with PB1: 84.\n",
      "Inferring reassortment with the NA segment...\n",
      "\tInferred reassortment events with NA: 51.\n",
      "Inferring reassortment with the NP segment...\n",
      "\tInferred reassortment events with NP: 118.\n",
      "Inferring reassortment with the PA segment...\n",
      "\tInferred reassortment events with PA: 145.\n",
      "Inferring reassortment with the MP segment...\n",
      "\tInferred reassortment events with MP: 98.\n",
      "\n",
      "Estimated reassortment rate per ancestral lineage per year.\n",
      "\tBernoulli estimate: 0.213007\n",
      "\tPoisson estimate (for dense datasets): 0.202527\n",
      "Saved the annotated tree file to trees_3/annotated.tree\n",
      "P-value threshold for significance was set to 0.001\n",
      "Read 7 segments: HA, PB2, PB1, NA, NP, PA, MP\n",
      "Estimating molecular clock rates for each segment (TreeTime)...\n",
      "\tHA estimated molecular clock rate: 0.0013003 (R^2 = 0.889)\n",
      "\tPB2 estimated molecular clock rate: 0.0016705 (R^2 = 0.934)\n",
      "\tPB1 estimated molecular clock rate: 0.0013672 (R^2 = 0.948)\n",
      "\tNA estimated molecular clock rate: 0.0012655 (R^2 = 0.914)\n",
      "\tNP estimated molecular clock rate: 0.0012612 (R^2 = 0.882)\n",
      "\tPA estimated molecular clock rate: 0.0014513 (R^2 = 0.917)\n",
      "\tMP estimated molecular clock rate: 0.0004999 (R^2 = 0.848)\n",
      "Found 627 strains in common across the alignments.\n",
      "Optimally resolving the multifurcations to minimize reassortment...\n",
      "Inferring reassortment with the PB2 segment...\n",
      "\tInferred reassortment events with PB2: 126.\n",
      "Inferring reassortment with the PB1 segment...\n",
      "\tInferred reassortment events with PB1: 85.\n",
      "Inferring reassortment with the NA segment...\n",
      "\tInferred reassortment events with NA: 54.\n",
      "Inferring reassortment with the NP segment...\n",
      "\tInferred reassortment events with NP: 117.\n",
      "Inferring reassortment with the PA segment...\n",
      "\tInferred reassortment events with PA: 146.\n",
      "Inferring reassortment with the MP segment...\n",
      "\tInferred reassortment events with MP: 96.\n",
      "\n",
      "Estimated reassortment rate per ancestral lineage per year.\n",
      "\tBernoulli estimate: 0.216922\n",
      "\tPoisson estimate (for dense datasets): 0.204165\n",
      "Saved the annotated tree file to trees_4/annotated.tree\n",
      "P-value threshold for significance was set to 0.001\n",
      "Read 7 segments: HA, PB2, PB1, NA, NP, PA, MP\n",
      "Estimating molecular clock rates for each segment (TreeTime)...\n",
      "\tHA estimated molecular clock rate: 0.0013003 (R^2 = 0.889)\n",
      "\tPB2 estimated molecular clock rate: 0.0016705 (R^2 = 0.934)\n",
      "\tPB1 estimated molecular clock rate: 0.0013672 (R^2 = 0.948)\n",
      "\tNA estimated molecular clock rate: 0.0012655 (R^2 = 0.914)\n",
      "\tNP estimated molecular clock rate: 0.0012612 (R^2 = 0.882)\n",
      "\tPA estimated molecular clock rate: 0.0014513 (R^2 = 0.917)\n",
      "\tMP estimated molecular clock rate: 0.0004999 (R^2 = 0.848)\n",
      "Found 627 strains in common across the alignments.\n",
      "Optimally resolving the multifurcations to minimize reassortment...\n",
      "Inferring reassortment with the PB2 segment...\n",
      "\tInferred reassortment events with PB2: 124.\n",
      "Inferring reassortment with the PB1 segment...\n",
      "\tInferred reassortment events with PB1: 84.\n",
      "Inferring reassortment with the NA segment...\n",
      "\tInferred reassortment events with NA: 49.\n",
      "Inferring reassortment with the NP segment...\n",
      "\tInferred reassortment events with NP: 118.\n",
      "Inferring reassortment with the PA segment...\n",
      "\tInferred reassortment events with PA: 143.\n",
      "Inferring reassortment with the MP segment...\n",
      "\tInferred reassortment events with MP: 96.\n",
      "\n",
      "Estimated reassortment rate per ancestral lineage per year.\n",
      "\tBernoulli estimate: 0.209295\n",
      "\tPoisson estimate (for dense datasets): 0.200344\n",
      "Saved the annotated tree file to trees_5/annotated.tree\n"
     ]
    }
   ],
   "source": [
    "list_of_genes = [\"ha\", \"pb2\",\"pb1\",\"na\",\"np\",\"pa\",\"mp\"]\n",
    "wrap_prep(list_of_genes, \n",
    "          \"preprepped\", \"aln\", \"metadata\", \"trees\", \n",
    "          \"prepped\", \"ha\", mincut =True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1276ff42-152e-43d2-bea0-7f005e35cd5c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
