{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "58d81ae5-353f-4279-8c5b-08e1575ff590",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0e149b39-daaa-49f8-8db3-5cd6c99ed348",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this converts fasta files into dataframes\n",
    "def fasta_to_df(fasta_file):\n",
    "    \n",
    "    fasta_data = []\n",
    "    \n",
    "    with open(fasta_file) as f:\n",
    "        header = \"\"\n",
    "        sequence = \"\"\n",
    "        for line in f:\n",
    "            if line.startswith(\">\"):\n",
    "                if header != \"\":\n",
    "                    fasta_data.append({\"header\": header, \"sequence\": sequence})\n",
    "                header = line.strip() \n",
    "                sequence = \"\"\n",
    "            else:\n",
    "                sequence += line.strip()\n",
    "        fasta_data.append({\"header\": header, \"sequence\": sequence}) #last line\n",
    "            \n",
    "    return pd.DataFrame(fasta_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "8bf89be8-a551-4381-a5eb-95e3f3fa4fef",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasta_writer(path, filename, df):\n",
    "            \n",
    "    try:  \n",
    "        os.mkdir(path)\n",
    "\n",
    "    except OSError as error:\n",
    "        pass\n",
    "\n",
    "    with open(f\"{path}{filename}\", \"w\") as f:\n",
    "        for index, row in df.iterrows():\n",
    "            f.write(f\"{row['header']}\\n\")\n",
    "            f.write(f\"{row['sequence']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "bb234aa4-45ad-406a-bbab-71d890ccacb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_genes = [\"ha\", \"pb1\", \"pb2\",\"pa\",\"mp\",\"np\",\"na\",\"ns\"]\n",
    "\n",
    "hosts = [\"allSwine\",\n",
    "         \"swineEuro\",\n",
    "         \"swineNA\",\n",
    "         \"allAvian\",\n",
    "         \"avianNA\",\n",
    "         \"avianEurasia\",\n",
    "         \"canineH3N8\",\n",
    "         \"equine\",\n",
    "         \"human\",\n",
    "         \"canineH3N2\"\n",
    "        ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6a4ea5e7-a635-4d6f-a5b2-65670018077c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def alignments(hosts: list, list_of_genes: list):\n",
    "\n",
    "    for host in hosts:\n",
    "\n",
    "        for gene in list_of_genes:\n",
    "\n",
    "            alignment = f\"./alignments/pruned_alignments_h3nx_{gene}.fasta\"\n",
    "\n",
    "            align = fasta_to_df(alignment)\n",
    "            align['header_stripped'] = align['header'].str.lstrip('>')\n",
    "\n",
    "            leaves_file = f\"./alignments/leaves/{host}/{host}_leaves_{gene}.tsv\"\n",
    "            leaves_df = pd.read_csv(leaves_file, header=None, names=[\"strain\"])\n",
    "\n",
    "            align_filtered = align[align['header_stripped'].isin(leaves_df['strain'])].copy()\n",
    "\n",
    "            align_filtered['header'] = '>' + align_filtered['header_stripped']\n",
    "\n",
    "            align_filtered = align_filtered.drop(columns=['header_stripped'])\n",
    "\n",
    "            print(host, align_filtered.shape) \n",
    "\n",
    "            fasta_writer(f\"./{host}/\", f\"aligned_h3nx_{gene}.fasta\", align_filtered)\n",
    "                   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "fbbeef7b-4f37-45ff-ab88-45aca22ee34a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "allSwine (1902, 2)\n",
      "allSwine (1940, 2)\n",
      "allSwine (1931, 2)\n",
      "allSwine (1925, 2)\n",
      "allSwine (1931, 2)\n",
      "allSwine (1933, 2)\n",
      "allSwine (1912, 2)\n",
      "allSwine (1938, 2)\n",
      "swineEuro (349, 2)\n",
      "swineEuro (363, 2)\n",
      "swineEuro (361, 2)\n",
      "swineEuro (361, 2)\n",
      "swineEuro (379, 2)\n",
      "swineEuro (362, 2)\n",
      "swineEuro (350, 2)\n",
      "swineEuro (362, 2)\n",
      "swineNA (646, 2)\n",
      "swineNA (840, 2)\n",
      "swineNA (838, 2)\n",
      "swineNA (841, 2)\n",
      "swineNA (253, 2)\n",
      "swineNA (841, 2)\n",
      "swineNA (716, 2)\n",
      "swineNA (838, 2)\n",
      "allAvian (2310, 2)\n",
      "allAvian (2317, 2)\n",
      "allAvian (2317, 2)\n",
      "allAvian (2317, 2)\n",
      "allAvian (2318, 2)\n",
      "allAvian (2318, 2)\n",
      "allAvian (2318, 2)\n",
      "allAvian (2316, 2)\n",
      "avianNA (1205, 2)\n",
      "avianNA (1340, 2)\n",
      "avianNA (1341, 2)\n",
      "avianNA (1336, 2)\n",
      "avianNA (1345, 2)\n",
      "avianNA (1273, 2)\n",
      "avianNA (861, 2)\n",
      "avianNA (505, 2)\n",
      "avianEurasia (861, 2)\n",
      "avianEurasia (854, 2)\n",
      "avianEurasia (858, 2)\n",
      "avianEurasia (849, 2)\n",
      "avianEurasia (855, 2)\n",
      "avianEurasia (863, 2)\n",
      "avianEurasia (525, 2)\n",
      "avianEurasia (639, 2)\n",
      "canineH3N8 (46, 2)\n",
      "canineH3N8 (46, 2)\n",
      "canineH3N8 (46, 2)\n",
      "canineH3N8 (46, 2)\n",
      "canineH3N8 (45, 2)\n",
      "canineH3N8 (46, 2)\n",
      "canineH3N8 (43, 2)\n",
      "canineH3N8 (43, 2)\n",
      "equine (224, 2)\n",
      "equine (224, 2)\n",
      "equine (224, 2)\n",
      "equine (224, 2)\n",
      "equine (226, 2)\n",
      "equine (226, 2)\n",
      "equine (226, 2)\n",
      "equine (226, 2)\n",
      "human (234, 2)\n",
      "human (221, 2)\n",
      "human (229, 2)\n",
      "human (230, 2)\n",
      "human (227, 2)\n",
      "human (229, 2)\n",
      "human (233, 2)\n",
      "human (228, 2)\n"
     ]
    }
   ],
   "source": [
    "alignments(hosts, list_of_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfcbf86f-e0a1-486b-b56d-eea1409b5a92",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
